src/core/Tm.ml

```diff
@@ -10,6 +10,37 @@ type bface = (tm, btm) Desc.Boundary.face
 type bsys = (tm, btm) Desc.Boundary.sys
 type data_desc = (tm, btm) Desc.desc
 
+type ('a, 'k) telescope =
+  | TNil of 'k
+  | TCons of 'a * ('a, 'k) telescope bnd
+
+
+module NewDesc =
+struct
+  type btm =
+    | Var of int
+    | Intro of
+        {clbl : string;
+         const_args : tm list;
+         rec_args : btm list;
+         rs : tm list}
+
+  type bface = tm * tm * btm
+  type bsys = bface list
+
+  type const_spec = [`Const of tm]
+  type rec_spec = [`Self]
+  type dim_spec = [`I]
+
+  type boundary_spec = bsys
+  type param_spec = [`Param of tm]
+
+  type constr = string * (const_spec, (rec_spec, (dim_spec, boundary_spec) telescope) telescope) telescope
```

Comment:
I agree, but I'm trying to write this code without needing to rewrite in a *conceptual* (i.e. not just formal) way the entire algorithms of HITs. This is to avoid getting screwed, trying to figure out which of the many changes I did at once introduced what bug.

After parametric HITs are implemented, I definitely encourage the change that you describe.

Reply:
In my experience, the code is almost the same (and arguably more systematic).

Reply:
@favonia I'm starting to see your point... Would you mind proposing a revised version of the `telescope` datatype with all the constructors you describe?

Reply:
The thing I don't quite grasp is, how can anything be dependent on the `Rec` parts? (I guess this is a question about both my version and your version.)

Reply:
I think the main challenge is to deal with the situation where we have the intermediate state of a datatype existing (i.e. some constructors, but no eliminators). I remember that there was question in RedPRL of how to do this---what did you end up doing?

Reply:
@jonsterling No arguments can depend on the `Rec` part, at least not right now, though boundaries can. However, the non-dependency will be enforced by type-checking automatically.

Reply:
> I think the main challenge is to deal with the situation where we have the intermediate state of a datatype existing (i.e. some constructors, but no eliminators). I remember that there was question in RedPRL of how to do this---what did you end up doing?

It was done by two passes. The first pass grabs the applicative structure of all constructors, and then the second pass is the real elaboration. The first pass recognizes the special keyword `self`. (I am not sure if I am answering your question.)

Reply:
@favonia The new more flexible schema that you propose kind of requires a global rewrite of all the HIT code, I think. I support the idea, but I think I would need some help to get it done.

Reply:
I think what I will start trying to do first is to rectify some really nasty and dumb aspects of the elaborator, which will make the task a bit easier.

Reply:
> @favonia The new more flexible schema that you propose kind of requires a global rewrite of all the HIT code, I think. I support the idea, but I think I would need some help to get it done.

I can take a look once you have done the parameters. I believe a rewrite will also close #257 automatically.

Reply:
Right now, I think I am so owned by stupid de bruijn shit that it might be simpler to do the rewrite *before* adding paremeters rather than the other way around. not sure.